runtime/src/main/java/org/corfudb/protocols/logprotocol/CheckpointEntry.java

```diff
@@ -0,0 +1,217 @@
+    };
+
+    @RequiredArgsConstructor
+    public enum CheckpointDictKey {
+        START_TIME(0),
+        END_TIME(1),
+        START_LOG_ADDRESS(2),
+        ENTRY_COUNT(3),
+        BYTE_COUNT(4);
+
+        public final int type;
+
+        public byte asByte() {
+            return (byte) type;
+        }
+
+        public static final Map<Byte, CheckpointDictKey> typeMap =
+                Arrays.stream(CheckpointDictKey.values())
+                        .collect(Collectors.toMap(CheckpointDictKey::asByte, Function.identity()));
+    }
+
+    /** Type of entry
+     */
+    @Getter
+    CheckpointEntryType cpType;
+
+    /**
+     * Unique identifier for this checkpoint.  All entries
+     * for the same checkpoint state must use the same ID.
+     */
+    @Getter
+    UUID checkpointID;
+
+    /** Author/cause/trigger of this checkpoint
+     */
+    @Getter
+    String checkpointAuthorID;
+
+    /** Map of checkpoint metadata, see key constants above
+     */
+    @Getter
+    Map<CheckpointDictKey, String> dict;
+
+    /** Optional: array of SMREntry objects that contains
+     *  SMR object state of the stream that we're checkpointing.
+     *  May be present in any CheckpointEntryType, but typically
+     *  used by CONTINUATION entries.
+     */
+    @Getter
+    @Setter
+    SMREntry[] smrEntries;
+
+    /** Byte count of smrEntries in serialized form, zero
+     *  if smrEntries.size() is zero or if value is unknown.
+     */
+    @Getter
+    int smrEntriesBytes = 0;
+
+    public CheckpointEntry(CheckpointEntryType type, String authorID, UUID checkpointID,
+                           Map<CheckpointDictKey,String> dict, SMREntry[] smrEntries) {
+        super(LogEntryType.SMR);
+        this.cpType = type;
+        this.checkpointID = checkpointID;
+        this.checkpointAuthorID = authorID;
+        this.dict = dict;
+        this.smrEntries = smrEntries;
+    }
+
+    /**
+     * This function provides the remaining buffer. Child entries
+     * should initialize their contents based on the buffer.
+     *
+     * @param b The remaining buffer.
+     * @param rt The CorfuRuntime used by the SMR object.
+     * @return A CheckpointEntry.
+     */
+    @Override
+    void deserializeBuffer(ByteBuf b, CorfuRuntime rt) {
+        super.deserializeBuffer(b, rt);
+        cpType = CheckpointEntryType.typeMap.get(b.readByte());
+        long cpidMSB = b.readLong();
+        long cpidLSB = b.readLong();
+        checkpointID = new UUID(cpidMSB, cpidLSB);
+        checkpointAuthorID = deserializeString(b);
+        dict = new HashMap<>();
+        short mapEntries = b.readShort();
+        for (short i = 0; i < mapEntries; i++) {
+            CheckpointDictKey k = CheckpointDictKey.typeMap.get(b.readByte());
+            String v = deserializeString(b);
+            dict.put(k, v);
+        }
+        smrEntries = null;
+        int items = b.readShort();
+        if (items > 0) {
```

Comment:
1 & 2: because I write C flavored Java at first draft.  I'll look into it.

3: The comment says where & why the error would happen but omits the NullPointerException detail.  Commit d1d019e makes this unnecessary now, I believe.  I'll remove it.